litho-processor/src/main/java/com/facebook/litho/specmodels/generator/EventGenerator.java

```diff
@@ -119,16 +124,103 @@ static TypeSpecDataHolder generateEventDispatcher(EventDeclarationModel eventDec
+    * */
+    if (eventClassHasNonEmptyCtor) {
+      final StringBuilder ctorParamsPlaceholder = new StringBuilder();
+      final List<String> ctorParamNames = new ArrayList<>();
+
+      /*
+      * Iterate over the Event fields in order to achieve the following:
+      *
+      * 1) Add the parameters to the method under construction.
+      * 2) Create a parameter name placeholder for JavaPoet.
+      * 3) Keep a list of all parameter names.
+      * */
+      for (FieldModel fieldModel : eventDeclaration.fields) {
+        eventDispatcherMethod
+            .addParameter(fieldModel.field.type, fieldModel.field.name);
+
+        // We do not want ',' appended on the first loop
+        if (!ctorParamNames.isEmpty()) {
+          ctorParamsPlaceholder.append(", ");
+        }
 
-    for (FieldModel fieldModel : eventDeclaration.fields) {
-      if (fieldModel.field.modifiers.contains(Modifier.FINAL)) {
-        continue;
+        ctorParamsPlaceholder.append("$L");
+
+        ctorParamNames.add(fieldModel.field.name);
       }
+
+      /*
+      * Create a list with all the desired arguments to be for the following statement.
+      * These need to be the following:
+      *
+      * 1) Event class name (type declaration)
+      * 2) Event class name (class instantiation)
+      * 3) Event class constructor parameter names.
+      * */
+      final List<Object> ctorStatementArgs = new ArrayList<>();
+      ctorStatementArgs.add(0, eventDeclaration.name);
+      ctorStatementArgs.add(1, eventDeclaration.name);
+      ctorStatementArgs.addAll(ctorParamNames);
+
+      eventDispatcherMethod
+          .addStatement(
+              "final $T _eventState = new $T(" + ctorParamsPlaceholder + ")",
+              ctorStatementArgs.toArray(new Object[0])
+          );
+    } else {
+      /*
+      * Fallback way in case our Event class has empty constructors, where we assume, that the
+      * user of the API is probably using mutable fields in order to mutate the Event class.
+      * */
+      final boolean isKotlinClass = SpecElementTypeDeterminator.determine(eventRepresentedObject)
+          == SpecElementType.KOTLIN_CLASS;
+
       eventDispatcherMethod
-          .addParameter(fieldModel.field.type, fieldModel.field.name)
-          .addStatement("_eventState.$L = $L", fieldModel.field.name, fieldModel.field.name);
+          .addStatement(
+              "final $T _eventState = new $T()",
+              eventDeclaration.name,
+              eventDeclaration.name
+          );
+
+      for (FieldModel fieldModel : eventDeclaration.fields) {
+        if (fieldModel.field.modifiers.contains(Modifier.FINAL)) {
+          continue;
+        }
+
+        final boolean fieldIsPrivate = fieldModel.field.modifiers.contains(Modifier.PRIVATE);
+
+        eventDispatcherMethod
+            .addParameter(fieldModel.field.type, fieldModel.field.name);
+
+        if (isKotlinClass && fieldIsPrivate) {
```

Comment:
I would be very tempted to not even support this and instead use the validation to provide a nice error message on how they should use a data class :) I'm very happy if we can have fewer ways of doing the same thing

Reply:
@astreet I was totally thinking of the validation in combination with this code's removal! Will take a look at it tomorrow and will add the proper validation plus removing this kind of "hack" for Kotlin!